### PR TITLE
[WFS provider] Fix recovery of GeoServer server error when lack of primary key (fixes #29844)

### LINKED ISSUE #29844: Error importing cover from Geoserver WFS service
Author Name: **Mauro preda** (Mauro preda)
Original Redmine Issue: [22030](https://issues.qgis.org/issues/22030)
Affected QGIS version: 3.4.6
Redmine category:web_services_clients/wfs

---
In geoservers (http://geoserver.org/) there are levels coming from postgis tables and views created with geoserver sql view.

Access to WFS works perfectly for all complete postgis covers .. but goes wrong if it's a cover created with sql vista

here is the error message:

2019-05-09T11:40:40     CRITICAL    Layer te_108_t : Scarico delle geometrie per il vettore sit_area_infrastrutture:te_108_t fallito o parzialmente fallito: Scarico delle geometrie non riuscito: Error transferring http://89.96.217.121:80/geoserver/wfs?SERVICE=WFS&REQUEST=GetFeature&VERSION=2.0.0&TYPENAMES=sit_area_infrastrutture:te_108_t&STARTINDEX=0&COUNT=1000000&SRSNAME=urn:ogc:def:crs:EPSG::32632&BBOX=469878.84158786124316975,5005658.55914706736803055,557148.479013089556247,5062663.48473474942147732,urn:ogc:def:crs:EPSG::32632 - server replied: . Puoi tentare di ricaricare il vettore con F5

Versione di QGIS
3.4.6-Madeira

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

File: src/providers/wfs/qgswfsfeatureiterator.cpp [hunk 1]
```diff
@@ -574,12 +569,18 @@ void QgsWFSFeatureDownloader::run( bool serializeFeatures, int maxFeatures )
...
-        QgsMessageLog::logMessage( mErrorMessage, tr( "WFS" ) );
+        if ( mErrorCode == NoError )
+        {
+          mErrorMessage = tr( "Error when parsing GetFeature response" ) + " : " + gmlProcessErrorMsg;
+          QgsMessageLog::logMessage( mErrorMessage, tr( "WFS" ) );
+        }
         break;
       }
-      if ( parser->isException() && finished )
+      if ( parser->isException() )
```
Comment (nyalldawson) [inline src/providers/wfs/qgswfsfeatureiterator.cpp]:
```suggestion
      else if ( parser->isException() )
```

File: src/providers/wfs/qgswfsfeatureiterator.cpp [hunk 2]
```diff
@@ -574,12 +569,18 @@ void QgsWFSFeatureDownloader::run( bool serializeFeatures, int maxFeatures )
       if ( !parser->processData( data, finished, gmlProcessErrorMsg ) )
       {
         success = false;
-        mErrorMessage = tr( "Error when parsing GetFeature response" ) + " : " + gmlProcessErrorMsg;
-        QgsMessageLog::logMessage( mErrorMessage, tr( "WFS" ) );
+        if ( mErrorCode == NoError )
+        {
+          mErrorMessage = tr( "Error when parsing GetFeature response" ) + " : " + gmlProcessErrorMsg;
```
Comment (nyalldawson) [inline src/providers/wfs/qgswfsfeatureiterator.cpp]: Took me a while to work out the logic here -- could you insert a comment describing the situation? I'm concerned someone might come along and think this is a typo, and change this to `mErrorCode != NoError`

File: src/providers/wfs/qgswfsfeatureiterator.cpp (hunk 1 above)
Comment (rouault) [reply src/providers/wfs/qgswfsfeatureiterator.cpp]: done

File: src/providers/wfs/qgswfsfeatureiterator.cpp (hunk 2 above)
Comment (rouault) [reply src/providers/wfs/qgswfsfeatureiterator.cpp]: done

**Pull request merged**